OpenRA.Mods.Common/Traits/Player/ParallelProductionQueue.cs

```diff
@@ -0,0 +1,70 @@
+﻿using System.Collections.Generic;
+
+namespace OpenRA.Mods.Common.Traits
+{
+	public class ParallelProductionQueueInfo : ProductionQueueInfo
+	{
+		public override object Create(ActorInitializer init) { return new ParallelProductionQueue(init, this); }
+	}
+
+	public class ParallelProductionQueue : ProductionQueue
+	{
+		[Sync] public new bool CurrentDone { get { return Queue.Find(i => i.Done) != null; } }
+
+		public ParallelProductionQueue(ActorInitializer init, ProductionQueueInfo info) : base(init, init.Self.Owner.PlayerActor, info)
+		{
+		}
+
+		protected override void ProgressQueue()
+		{
+			for (var i = 0; i < Queue.Count; i++)
+			{
+				var first = Queue[0];
+
+				var before = first.RemainingTime;
+				first.Tick(playerResources);
+
+				foreach (var item in Queue.FindAll(a => a.Item == first.Item))
+				{
+					Queue.Remove(item);
+					Queue.Add(item);
+				}
+
+				if (first.RemainingTime != before)
+					return;
+			}
+		}
+
+		protected override void PauseProduction(string itemName, uint extraData)
+		{
+			var item = Queue.Find(a => a.Item == itemName);
+
+			if (item != null)
+				item.Pause(extraData != 0);
+		}
+
+		public override bool IsProducing(ProductionItem item)
+		{
+			return Queue.Contains(item);
+		}
+
+		public override int RemainingTimeActual(ProductionItem item)
+		{
+			var parallelBuilds = 0;
+			var added = new List<string>();
+
+			foreach (var i in Queue)
+			{
+				if (added.Contains(i.Item))
+					continue;
+
+				added.Add(i.Item);
+
+				if (!i.Paused && i.RemainingTime > 0)
+					parallelBuilds++;
+			}
+
+			return item.RemainingTimeActual * parallelBuilds;
```

Comment:
Am I misunderstanding something? Because the point of a build being parallel is they do something at the same time so to multiply the time with the number of builds looks weird to me.

Reply:
If you build multiple units at the same time without additional time you get a huge balancing problem. in the end you will construct anything at the same time possible, as it enlarges your army. The result should be that the time it takes to build two different actors will be the same. However this allows you to build an harvester or a fairly equal army of 2-3 units at the same time without the need to cancel everything else.

Reply:
imagine a tank of 100 ticks and a harvester of 50 ticks, and imagine 100 ticks takes a very long time ;) now your tank is already 50% complete. and you need an harvester too. either you cancel the tank and build the harvester in 50 ticks, trashing the 50 ticks time for the tank or you wait 50 ticks, build the harvester, wait another 50 ticks and you got it in 100 ticks. using parallel build both will take 100 more ticks as producing two things makes it take twice as long. so in the end both are done at the same time, and you did not waste 50ticks for the tank ;)